[sonic-mgmt] Update transceiver onboarding HLD to make parameters attribute based

mihirpat1 opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->
This PR modernizes and significantly enhances the transceiver onboarding test plan by introducing a robust, modular attribute management system that replaces the previous CSV-based approach with a comprehensive JSON-based configuration framework.

Also, moved the `transceiver_onboarding_test_plan.md` file to `transceiver/test_plan.md`.

**PRs planned to add attributes for various test categories**
| PR title | state | 
| ---------------------------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------- |
| [Move Transceiver Onboarding Test Plan to transceiver folder](https://github.com/sonic-net/sonic-mgmt/pull/21100) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-mgmt/21100) |
| [Transceiver firmware upgrade testplan](https://github.com/sonic-net/sonic-mgmt/pull/20102) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-mgmt/20102) |
|[[Transceiver onboarding] Add EEPROM and System test attributes in HLD](https://github.com/sonic-net/sonic-mgmt/pull/20359)| ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-mgmt/20359) |
| [Optics insertion removal testplan](https://github.com/sonic-net/sonic-mgmt/pull/20524) | ![GitHub issue/pull request detail](https://img.shields.io/github/pulls/detail/state/Azure/sonic-mgmt/20524) |

Summary:
Fixes # (issue)

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [x] Testbed and Framework(new/improvement)
- [ ] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

### Approach
#### What is the motivation for this PR?

#### How did you do it?

#### How did you verify/test it?

#### Any platform specific information?

#### Supported testbed topology if it's a new test case?

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

ADO - 33984323

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

az-pz reviewed (commented):
Minor comments. Looks great otherwise. Very thorough and comprehensive.

mssonicbld commented:
/azp run

az-pz commented on docs/testplan/transceiver_onboarding_test_plan.md:
```diff
@@ -116,23 +118,37 @@ These tests do not require traffic and are standalone, designed to run on a Devi
-        "port_1": {
-          "vendor_name": "Vendor/Inc",
-          "vendor_pn": "QSFP-100G-AOC-10M"
+        "Ethernet0-96:4": {
```
Should we use Python list slicing syntax instead?

mihirpat1 replied:
@az-pz I have addressed this now

mssonicbld commented:
/azp run

az-pz commented on docs/testplan/transceiver_onboarding_test_plan.md:
```diff
@@ -144,6 +160,34 @@ These tests do not require traffic and are standalone, designed to run on a Devi
 
+    **Port Configuration Keys:**
+    - **Individual Ports**: Standard SONiC port names (e.g., `"Ethernet0"`, `"Ethernet4"`)
+    - **Port Ranges**: Range notation (e.g., `"Ethernet4-12"`, `"Ethernet0-96:4"`)
```
`"Ethernet0-96:4"` should be updated to `"Ethernet0:97:4"`.

mssonicbld commented:
/azp run

az-pz reviewed (approved):
Very thorough and comprehensive. Looks great.

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -98,87 +98,593 @@ These tests do not require traffic and are standalone, designed to run on a Devi
 **Pre-requisites for the Below Tests:**
 
-1. A file `transceiver_dut_info.csv` (located in `ansible/files/transceiver_inventory` directory) should be present to describe the metadata of the transceiver connected to every port of each DUT. Following should be the format of the file
+1. **DUT Info File:** A file `transceiver_dut_info.json` (located in `ansible/files/transceiver/inventory` directory) should be present to describe the metadata of the transceiver connected to corresponding ports of each DUT. This file serves as the foundational source of truth for transceiver identification and contains data that is unique to each physical transceiver and its corresponding port  ...
```
Ensured that all paths are consistent

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -98,87 +98,593 @@ These tests do not require traffic and are standalone, designed to run on a Devi
+    **Field Definitions:**
+
+    **Global Normalization Mappings:**
+    - `normalization_mappings.vendor_names`: Dictionary mapping raw vendor names to their normalized forms using the normalization rules described in the [CMIS CDB Firmware Binary Management](#141-cmis-cdb-firmware-binary-management) section.
+    - `normalization_mappings.part_numbers`: Dictionary mapping raw part numbers to their normalized forms using the normalization rules described in the [CMIS CDB Firmware Binary Management](#141-cmis-cdb-firmware-binary-management) section.
```
This section currently exists

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -98,87 +98,593 @@ These tests do not require traffic and are standalone, designed to run on a Devi
+            return {}
+        
+        parts = config_string.split('-')
+        if len(parts) != 6:
+            raise ValueError(f"Invalid transceiver configuration format: {config_string}")
```
Addressed this now.

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -225,7 +731,8 @@ The following tests aim to validate the link status and stability of transceiver
 | Issue CLI command to shutdown a port | Validate link status using CLI configuration | Ensure that the link goes down |
 | Issue CLI command to startup a port | Validate link status using CLI configuration | Ensure that the link is up and the port appears in the LLDP table. |
-| In a loop, issue startup/shutdown command 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
+| In a loop, issue startup/shutdown command for a port 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
+| In a loop, issue startup/shutdown command for all ports 100 times | Stress test for link status validation | Ensure link status toggles to up/down appropriately for all relevant ports with each startup/shutdown command. Verify ports appear in the LLDP table when the link is up |
```
This is being addressed via https://github.com/sonic-net/sonic-mgmt/pull/20359

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -0,0 +1,1448 @@
+| Restart `swss` | Validate transceiver re-initialization and link status post container restart | Ensure `xcvrd` restarts (for Mellanox platform, ensure pmon restarts) and the expected ports link up again, with port details visible in the LLDP table |
+| Restart `syncd` | Validate transceiver re-initialization and link status post container restart | Ensure `xcvrd` restarts (for Mellanox platform, ensure pmon restarts) and the expected ports link up again, with port details visible in the LLDP table |
+| Perform a config reload | Test transceiver re-initialization and link status | Ensure `xcvrd` restarts and the expected ports link up again, with port details visible in the LLDP table |
+| Execute a cold reboot | Validate transceiver re-initialization and link status post-device reboot | Confirm the expected ports link up again post-reboot, with port details visible in the LLDP table |
```
@gshemesh2 Yes - I have added `cold_reboot_settle_sec` attribute to system attribute dictionary

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -0,0 +1,1448 @@
+| Restart `syncd` | Validate transceiver re-initialization and link status post container restart | Ensure `xcvrd` restarts (for Mellanox platform, ensure pmon restarts) and the expected ports link up again, with port details visible in the LLDP table |
+| Perform a config reload | Test transceiver re-initialization and link status | Ensure `xcvrd` restarts and the expected ports link up again, with port details visible in the LLDP table |
+| Execute a cold reboot | Validate transceiver re-initialization and link status post-device reboot | Confirm the expected ports link up again post-reboot, with port details visible in the LLDP table |
+| In a loop, execute cold reboot 100 times | Stress test to validate transceiver re-initialization and link status with cold reboot | Confirm the expected ports link up again post-reboot, with port details visible in the LLDP table |
```
This is configurable via `cold_reboot_iterations` attribute. Doing such soaking tests helps to test the system stability.

mssonicbld commented:
/azp run

prgeor commented on docs/testplan/transceiver/diagrams/data_flow.md:
````diff
@@ -0,0 +1,199 @@
+Result: dual_bank_supported = true
+```
+
+### Step 4: Final Structure
````
@mihirpat1 did we add the debug log which can be turned ON when necessary to see this final structure?

prgeor commented on docs/testplan/transceiver/diagrams/file_organization.md:
```diff
@@ -0,0 +1,68 @@
+├── prerequisites.json                         # OPTIONAL grouped prerequisite tests (pre-category gating)
+│
+└── templates/                              # Validation templates (optional)
+    └── deployment_templates.json           # Attribute completeness validation
```
@mihirpat1 How do we bind rest of the attributes to a particular deployment?

prgeor commented on docs/testplan/transceiver/diagrams/file_organization.md:
```diff
@@ -0,0 +1,68 @@
+ansible/files/transceiver/inventory/
+├── dut_info.json                           # Core transceiver metadata
+│
+├── attributes/                             # Test category attribute files
```
@mihirpat1 as discussed lets make the changes
1. Add sample dut_info as examples
2. DUT info that contains per port information should be in dut specific file matching the dut name so that over time this does not grow to huge size

mihirpat1 replied:
@prgeor I have addressed this now.

arpit-nexthop commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -96,126 +96,797 @@ These tests do not require traffic and are standalone, designed to run on a Devi
+
+**Requirements**:
+
+- Parse the mandatory 6-component format: `{TYPE}-{SPEED}-{FORM_FACTOR}-{DEPLOYMENT}-{MEDIA_LANE_MASK}-{HOST_LANE_MASK}`
```
Any thoughts on making this attribute a dictionary?

`"transceiver_configuration": "AOC-100-QSFP-100G_STRAIGHT-0x0F-0x0F"`

The string is quite lengthy and more often it will be generated, so we will be generating the configuration and then decoding it, it will be more intutive if the fields of this string are a dictionary.

arpit-nexthop replied:
@mihirpat1 I am more concerned with human understandability here than the file size.
`AOC-100-QSFP-100G_STRAIGHT-0x0F-0x0F` compresses lot of data that needs to be understood.

Just a preference on my part. I will let others comment on this as well.

mihirpat1 committed:
Refactor transceiver test plan: Add attribute-based infrastructure

- Restructure test plan into modular, attribute-based framework
- Introduce per-DUT configuration files in dut_info/ directory
- Add centralized normalization_mappings.json for vendor/PN normalization
- Implement category-specific attribute files (eeprom, system, etc.)
- Add deployment-based configuration grouping
- Support flexible port specifications (ranges, lists, mixed formats)
- Implement 8-level priority-based attribute resolution hierarchy
- Add prerequisite tests for early readiness validation
- Introduce optional attribute completeness validation with templates
- Create comprehensive examples in examples/inventory/ directory
- Add visual documentation (data flow, file organization, validation flow)
- Support platform/HWSKU-specific overrides
- Document transceiver configuration format with 6-component naming
- Enable selective test execution based on file presence

This refactoring improves scalability, maintainability, and separation of
concerns for transceiver onboarding test infrastructure.

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mihirpat1 committed:
Removed unwanted changed in transceiver_onboarding_test_plan.md

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mssonicbld commented:
/azp run

mihirpat1 committed:
Added pre-req to validate port speed and RS FEC

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mihirpat1 committed:
Added dut_hostname to dut_info filename

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mssonicbld commented:
/azp run

mihirpat1 committed:
Added test file hierarchy

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

prgeor commented on docs/testplan/transceiver/examples/inventory/attributes/eeprom.json:
```diff
@@ -0,0 +1,42 @@
+    "mandatory": ["sff8024_identifier", "dual_bank_supported"],
+    "defaults": {
+        "vdm_supported": false,
+        "is_non_dac_and_cmis": false,
```
@mihirpat1 I am not sure what this flag mean.

mihirpat1 replied:
@prgeor Following is the meaning of this
<img width="1502" height="85" alt="image" src="https://github.com/user-attachments/assets/f8b2da95-64e7-49e4-9daa-869a001aba49" />

This flag should be true for all CMIS optics which are non-DAC (400ZR, DR8 etc)

prgeor commented on docs/testplan/transceiver/examples/inventory/attributes/eeprom.json:
@mihirpat1 some optics may support few DOM flags and some not. how are we handling such scenario?

prgeor commented on docs/testplan/transceiver/test_plan.md:
@mihirpat1 should we limit only the infra part in this file? All detailed specific test should be in https://github.com/sonic-net/sonic-mgmt/pull/20359/changes ?

mihirpat1 committed:
Correct typo to eeprom_dump_timeout_sec

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mihirpat1 commented on docs/testplan/transceiver/examples/inventory/attributes/eeprom.json:
```diff
@@ -0,0 +1,42 @@
+    "defaults": {
+        "vdm_supported": false,
+        "is_non_dac_and_cmis": false,
+        "sfputil_eeprom_dump_time": 5
```
@prgeor Corrected this now to `eeprom_dump_timeout_sec`

mihirpat1 committed:
Updated file hierarchy to add more test group categories

Signed-off-by: Mihir Patel <patelmi@microsoft.com>

mssonicbld commented:
/azp run

mihirpat1 commented on docs/testplan/transceiver/test_plan.md:
```diff
@@ -431,13 +1127,13 @@ This section describes the automated process for copying firmware binaries to th
 To ensure only the necessary firmware binaries are present for each transceiver:
 
 1. **Parse `transceiver_firmware_info.csv`** to obtain the list of available firmware binaries, their versions, and associated vendor and part numbers.
```
@roy-sror We are planning to remove this section via https://github.com/sonic-net/sonic-mgmt/pull/20102 after the current PR is merged.
Will let @az-pz convert the CSV to JSON file via the above PR.
